Block API: Refactor the parser module in preparation for further work

dmsnell opened the pull request:
## What

This patch contains a mixture of small refators, mostly a combination of
updating comments and explanations as well as rearranging some array
iterating code to use more traditional loop control structures.

While the comments should be self-explanatory, the loop control
structures are refactored for a dual purpose. In framework code like
this that is core to the project and often untouched, there is value in
optimizing to eliminate memory allocations and reduce pressure on the
garbage collector. These costs are often difficult to profile in
existing tools. More imporantly, as we think about introducing an
asynchronous block loading system we need to prepare to handle the
semantic changes that asynchronous flow introduces.

These changes are meant to smooth the transition to asynchronous code
while clarifying how blocks are loaded into the editor.

## Why

It would be helpful to be able to load blocks asynchronously and defer
the loading of deprecation code and heavier block implementation code.
Right now that's not possible given that the block loading system is
fully synchronous.

This is the start of several changes that would be necessary to support
asynchronous loading, and this exists as a small piece of that which is
designed to be minimal in order to be easy to review.

## Testing

Since the changes in this PR impact fundamental data flow in the editor
the testing suites should catch most notable bugs. Please audit the code
and look for ways that subtle bugs could creep in, how the refactor may
not be true to the code it replaced.

hypest commented:
👋 Dennis! Is there any update on this PR? Is it perhaps in a state that someone can move it forward? Thanks!

dmsnell commented:
@hypest I haven't touched this in a long time. I would imagine it's easier to start over and recreate this to get it ready for merging (can re-use the PR and force a new branch `HEAD`).

I'm pretty occupied for a while. Are you wanting to work on it? I'd love to help if you do.

gziolo commented:
@dmsnell, is it something we still want to work on?

> In preparation for some work to refactor the handling of unsupported blocks in Gutenberg this patch is rearranging code in the block API parser module in order to simplify the logic of hooking up parsed blocks to their matched block implementation.

Some aspects of the initial proposal should be already resolved in alternative PRs.

dmsnell commented:
> is it something we still want to work on?

Yes! @gziolo I would love to come back to this and hopefully in the next week can come back to revive the PR.

 - I'd like to lock in the stage-1 and stage-2 terminology.
 - Some of the simplifications still apply.
 - This work is also prep-work for some asynchronous block-loading flows.

gziolo commented:
>  I would love to come back to this and hopefully in the next week can come back to revive the PR.

Awesome, I will watch the progress 😄

dmsnell committed:
Block API: Refactor parser module in preparation for future work.

This patch contains a mixture of small refators, mostly a combination of
updating comments and explanations as well as rearranging some array
iterating code to use more traditional loop control structures.

While the comments should be self-explanatory, the loop control
structures are refactored for a dual purpose. In framework code like
this that is core to the project and often untouched, there is value in
optimizing to eliminate memory allocations and reduce pressure on the
garbage collector. These costs are often difficult to profile in
existing tools. More imporantly, as we think about introducing an
asynchronous block loading system we need to prepare to handle the
semantic changes that asynchronous flow introduces.

These changes are meant to smooth the transition to asynchronous code
while clarifying how blocks are loaded into the editor.

dmsnell commented:
Rebased from 4979cbf28f0142dea8c73516a23dc2b013262bba to 62340740df2

youknowriad commented:
Doing some repository cleanup today, I was wondering if this PR is still relevant and worth keeping open?

dmsnell commented:
> Doing some repository cleanup today, I was wondering if this PR is still relevant and worth keeping open?

@youknowriad it would take some updating but I still find value in it because the parser still IMO lacks some clarifying vocabulary. I was curious how much it bumps into the work that @jsnajdr recently did.

jsnajdr commented:
The one part of the diff that looks familar to me are the changes to `parseRawBlock` and `parse` function that prepare them for being async by replacing functional `.map` and `.reduce` loops by a simple `for`.

My block lazy loading experiment does very similar changes here:
https://github.com/WordPress/gutenberg/pull/51778/files#diff-3629752a6442e4d3b18f333cd344e2d51128525fc5c921a90bcca1e1f5d7f591L226-L317

I don't immediately understand the changes in `convert-legacy-block.js`, I never had to make any changes there.

I also see a plenty of jsdoc changes and new jsdoc documentation. It could be valuable to extract these new docs and land them.
